routing/index_graph_loader.cpp

```diff
@@ -0,0 +1,107 @@
+
+#include "coding/file_container.hpp"
+
+#include "base/assert.hpp"
+
+namespace
+{
+using namespace routing;
+
+class IndexGraphLoaderImpl final : public IndexGraphLoader
+{
+public:
+  IndexGraphLoaderImpl(Index & index, shared_ptr<NumMwmIds> numMwmIds,
+                       shared_ptr<VehicleModelFactory> vehicleModelFactory,
+                       shared_ptr<EdgeEstimator> estimator, shared_ptr<TrafficStash> trafficStash);
+
+  // IndexGraphLoader overrides:
+  virtual IndexGraph & GetIndexGraph(NumMwmId numMwmId) override;
+
+private:
+  void Load(NumMwmId mwmId);
+
+  Index & m_index;
+  shared_ptr<NumMwmIds> m_numMwmIds;
+  shared_ptr<VehicleModelFactory> m_vehicleModelFactory;
+  shared_ptr<EdgeEstimator> m_estimator;
+  shared_ptr<TrafficStash> m_trafficStash;
+  unordered_map<NumMwmId, unique_ptr<IndexGraph>> m_graphs;
+};
+
+IndexGraphLoaderImpl::IndexGraphLoaderImpl(Index & index, shared_ptr<NumMwmIds> numMwmIds,
+                                           shared_ptr<VehicleModelFactory> vehicleModelFactory,
+                                           shared_ptr<EdgeEstimator> estimator,
+                                           shared_ptr<TrafficStash> trafficStash)
+  : m_index(index)
+  , m_numMwmIds(numMwmIds)
+  , m_vehicleModelFactory(vehicleModelFactory)
+  , m_estimator(estimator)
+  , m_trafficStash(trafficStash)
+{
+  CHECK(m_numMwmIds, ());
+  CHECK(m_vehicleModelFactory, ());
+  CHECK(m_estimator, ());
+  CHECK(m_trafficStash, ());
+}
+
+IndexGraph & IndexGraphLoaderImpl::GetIndexGraph(NumMwmId numMwmId)
+{
+  auto it = m_graphs.find(numMwmId);
+  if (it != m_graphs.end())
+    return *it->second;
+
+  Load(numMwmId);
+
+  it = m_graphs.find(numMwmId);
+  CHECK(it != m_graphs.end(), ());
+  return *it->second;
+}
+
+void IndexGraphLoaderImpl::Load(NumMwmId numMwmId)
+{
+  platform::CountryFile const & file = m_numMwmIds->GetFile(numMwmId);
+  MwmSet::MwmHandle handle = m_index.GetMwmHandleByCountryFile(file);
+  if (!handle.IsAlive())
+    MYTHROW(RoutingException, ("Can't get mwm handle for", file));
+
+  shared_ptr<IVehicleModel> vehicleModel =
+      m_vehicleModelFactory->GetVehicleModelForCountry(file.GetName());
+
+  auto const mwmId = MwmSet::MwmId(handle.GetInfo());
+  auto graph =
+      make_unique<IndexGraph>(GeometryLoader::Create(m_index, mwmId, vehicleModel), m_estimator);
+  m_trafficStash->Load(mwmId, numMwmId);
+
+  MwmValue const & mwmValue = *handle.GetValue<MwmValue>();
+
+  my::Timer timer;
+  FilesContainerR::TReader reader(mwmValue.m_cont.GetReader(ROUTING_FILE_TAG));
+  ReaderSource<FilesContainerR::TReader> src(reader);
+  IndexGraphSerializer::Deserialize(*graph, src, kCarMask);
+  RestrictionLoader restrictionLoader(mwmValue, *graph);
+  if (restrictionLoader.HasRestrictions())
+    graph->SetRestrictions(restrictionLoader.StealRestrictions());
+
+  m_graphs[numMwmId] = move(graph);
+  LOG(LINFO, (ROUTING_FILE_TAG, "section for", file.GetName(), "loaded in", timer.ElapsedSeconds(),
```

Comment:
Do we really need this in INFO? Am I right that this message will be displayed for each mwm needed for route construction, each time route is constructed?